### PR TITLE
Initial Rust AST implementation

### PR DESCRIPTION
**Pull request by mwu-tow**
### Pull Request Description
This PR adds initial Rust AST implementation. Due to technical reasons, this contains #347 — if you already reviewed it, skip ` common/rust/shapely` subtree.

AST lives in `ast` package, that relies significantly on `ast-macros` to generate boilerplate.
Additional `macro-utils` library was split out from `shapely-macros` and `ast-macros`.

The implementation was contributed by @wdanilo , I basically just did some refactoring, documenting and testing.

### Important Notes
* AST is known to be incomplete structurally, finishing it is #336 
* AST is missing a number of necessary functions, some of them explicitly marked as FIXME in the code, finishing them is #338 
* while I have written some tests, they are not yet part of CI — I want to this smart way (i.e. allowing tests that rely on parser), it is tracked as #340 
* AST JSON serialization is incompatible with Scala, solving this is #297 
* there is some non-deterministic issue with CI on Windows — I need to look into this closer but it seems to not be related to any Rust parts

### Checklist
Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [ ] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [ ] All code has been tested where possible.

### CODE REVIEW

File: common/rust/ast/core/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,19 @@
+[package]
+name    = "ast"
+version = "0.1.0"
+authors = ["Wojciech Danilo <wojciech.danilo@luna-lang.org>"]
```
Comment (wdanilo) [inline common/rust/ast/core/Cargo.toml]: Lets change it to Enso Team

File: common/rust/ast/macros/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+[package]
+name    = "ast-macros"
+version = "0.1.0"
+authors = ["Wojciech Danilo <wojciech.danilo@luna-lang.org>"]
```
Comment (wdanilo) [inline common/rust/ast/macros/Cargo.toml]: Enso Team

File: common/rust/ast/macros/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+Inflector      = "0.11.4"
+
+macro-utils    = { version = "0.1.0"  , path = "../../macro-utils" }
+prelude        = { version = "0.1.0"  , path = "../../prelude" }
+
+[dependencies.syn]
+version = "1.0"
+features = [
+  'extra-traits',
+  'full' # for syn::File and syn::ItemFn
```
Comment (wdanilo) [inline common/rust/ast/macros/Cargo.toml]: amazing comment ❤️

File: common/rust/ast/macros/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+// ==============
+// === Macros ===
+// ==============
+
+/// A macro that shall be applied to all AST nodes.
+///
+/// Derives all the traits that are expected to be implemented by AST nodes.
+///
+/// Implicitly applied by `ast` on target and generated types. User should not
+/// need to use this macro directly.
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: Why these empty lines in docs?

File: common/rust/ast/macros/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+#[proc_macro_attribute]
+pub fn ast_node
+(_meta: proc_macro::TokenStream
+ , input: proc_macro::TokenStream
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: very very minor - unnecessary space

File: common/rust/ast/macros/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+    output.into()
+}
+
+
+
+// ==============
+// === Macros ===
+// ==============
+
+/// Visitor that accumulates all visited `syn::TypePath` node representations.
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: Over 80 chars

File: common/rust/ast/macros/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+    let mut type_gather = TypeGatherer::new();
+    type_gather.visit_type(node);
+    type_gather.types
+}
+
+/// Produces declaration of the structure for given source enum variant.
+fn mk_product_type
+( is_flat : bool
+  , decl    : &syn::DeriveInput
+  , variant : &syn::Variant
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: spaces

File: common/rust/ast/macros/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+// Note [Expansion Example]
+// ~~~~~~~~~~~~~~~~~~~~~~~~
+// In order to make the definition easier to read, an example expansion of the
+// following definition was provided for each quotation:
+//
+// #[to_variant_types]
+// pub enum Shape<T> {
+//     Var(Var),
+//     App(App<T>),
+// }
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: I would put that node on top of this section - directly after section title "Macros"

File: common/rust/ast/macros/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        syn::Data::Enum(ref data) => data.variants.iter(),
+        _ => unimplemented!()
+    }.collect::<Vec<_>>();
+
+    let is_flat        = repr(&attrs) == "flat";
+    let variant_idents = variants.iter().map(|v| &v.ident).collect::<Vec<_>>();
+    let structs        = variants.iter().map(|v| mk_product_type(is_flat, &decl, v));
+    let structs        = structs.collect::<Vec<_>>();
+    let variant_decls  = variant_idents.iter().zip(structs.iter()).map(|(i,v)| gen_variant_decl(i,&v));
+    let variant_froms  = variant_idents.iter().zip(structs.iter()).map(|(i,v)| gen_from_impls(i, &decl, &v));
```
Comment (wdanilo) [inline common/rust/ast/macros/src/lib.rs]: Over 80 chars

File: common/rust/shapely/core/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+[package]
+name    = "shapely"
+version = "0.1.0"
+authors = ["Wojciech Danilo <wojciech.danilo@luna-lang.org>"]
```
Comment (wdanilo) [inline common/rust/shapely/core/Cargo.toml]: Enso Team

File: common/rust/shapely/macros/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,26 @@
+[package]
+name    = "shapely-macros"
+version = "0.1.0"
+authors = ["Wojciech Danilo <wojciech.danilo@luna-lang.org>"]
```
Comment (wdanilo) [inline common/rust/shapely/macros/Cargo.toml]: Enso Team

File: common/rust/ast/core/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,562 @@
...
+
+// === AST ===
+
+impl Ast {
+    // TODO smart constructors for other cases
+    //  as part of https://github.com/luna/enso/issues/338
+    pub fn var(name: String) -> Ast {
+        let var = Var{ name };
+        Ast::from(var)
+    }
```
Comment (wdanilo) [inline common/rust/ast/core/src/lib.rs]: I would like more generic approach. Why don't we return a generic type `T: From<Var>`? This way you can use `var` to generate both `Ast` as well as `Var` if needed! :)

File: common/rust/ast/core/src/lib.rs [hunk 13]
```diff
@@ -0,0 +1,562 @@
...
+    pub fn iter(&self) -> Box<dyn Iterator<Item = &'_ T> + '_> {
+        // TODO: use child's derived iterator
+        //  as part of https://github.com/luna/enso/issues/338
+        unimplemented!()
+    }
+}
+
+impl<T> HasSpan for Shape<T> {
+    // TODO: sum spans of all members
+    //  as part of https://github.com/luna/enso/issues/338
```
Comment (wdanilo) [inline common/rust/ast/core/src/lib.rs]: Just a note here - this should be a generic macro allowing us to just map "a function" over each variant child.

**Comment by wdanilo** [comment]
Very minor comments above.
Big applause for amazing tests 👏 

Approved, please see the comments and there is a green light for a merge!

File: common/rust/ast/core/Cargo.toml (hunk 1 above)
Comment (iamrecursion) [inline common/rust/ast/core/Cargo.toml]:
```suggestion
authors = ["Enso Team <contact@luna-lang.org>"]
```

File: common/rust/ast/macros/Cargo.toml (hunk 2 above)
Comment (iamrecursion) [inline common/rust/ast/macros/Cargo.toml]:
```suggestion
authors = ["Enso Team <contact@luna-lang.org>"]
```

File: common/rust/macro-utils/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,19 @@
+[package]
+name = "macro-utils"
+version = "0.1.0"
+authors = ["Wojciech Danilo <wojciech.danilo@luna-lang.org>"]
```
Comment (iamrecursion) [inline common/rust/macro-utils/Cargo.toml]:
```suggestion
authors = ["Enso Team <contact@luna-lang.org>"]
```

File: common/rust/macro-utils/Cargo.toml [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
```
Comment (iamrecursion) [inline common/rust/macro-utils/Cargo.toml]: Does this line need to be present?

File: common/rust/shapely/core/Cargo.toml (hunk 10 above)
Comment (iamrecursion) [inline common/rust/shapely/core/Cargo.toml]:
```suggestion
authors = ["Enso Team <contact@luna-lang.org>"]
```

File: common/rust/shapely/macros/Cargo.toml (hunk 11 above)
Comment (iamrecursion) [inline common/rust/shapely/macros/Cargo.toml]:
```suggestion
authors = ["Enso Team <contact@luna-lang.org>"]
```

File: common/rust/ast/macros/src/lib.rs (hunk 8 above)
Comment (mwu-tow) [reply common/rust/ast/macros/src/lib.rs]: done

File: common/rust/ast/macros/src/lib.rs (hunk 9 above)
Comment (mwu-tow) [reply common/rust/ast/macros/src/lib.rs]: I broke the lines, though I feel it should be done in a better way. But whatever I thought of, it requires names that are too long to solve the issue.

**Pull request merged**